### PR TITLE
[SYCL] Reduce overhead on command creation in specific case

### PR DESCRIPTION
**Pull request by alexanderfle**
The changes make even more performance optimization for calling kernels that have no dependencies, starting in the https://github.com/intel/llvm/pull/4188. Since the specific kernel doesn't change the execution command graph then the time spent on allocation/deallocation objects of the `ExecCGCommand` and `CGExecKernel` classes can be reduced by simply not creating them and directly calling kernel submission function(i.e `enqueueImpKernel`).

`case CG::CGTYPE::Kernel` from `ExecCGCommand::enqueueImp()` took a little refactoring to extract the code into a separate function - `enqueueImpKernel`, which can be called without creating the command.

Signed-off-by: Alexander Flegontov <alexander.flegontov@intel.com>

### CODE REVIEW

File: sycl/source/handler.cpp [hunk 1]
```diff
@@ -132,8 +136,47 @@ event handler::finalize() {
     }
   }
 
+  const auto &type = getType();
```
Comment (romanovvlad) [inline sycl/source/handler.cpp:139]: Could you please add more comments in this snippet?

File: sycl/source/detail/scheduler/commands.cpp [hunk 2]
```diff
@@ -1886,6 +1887,97 @@ void DispatchNativeKernel(void *Blob) {
     delete HostTask;
 }
 
+cl_int enqueueImpKernel(
+    const QueueImplPtr &Queue, NDRDescT &NDRDesc, std::vector<ArgDesc> &Args,
+    const std::shared_ptr<detail::kernel_bundle_impl> &KernelBundleImplPtr,
+    const std::shared_ptr<detail::kernel_impl> &MSyclKernel,
+    const std::string &KernelName, const detail::OSModuleHandle &OSModuleHandle,
+    std::vector<RT::PiEvent> &RawEvents, const EventImplPtr &EventImpl,
+    std::function<void *(Requirement *Req)> getMemAllocationFunc) {
```
Comment (romanovvlad) [inline sycl/source/detail/scheduler/commands.cpp:1896]:
```suggestion
    const std::function<void *(Requirement *Req)> &getMemAllocationFunc) {
```

File: sycl/source/detail/scheduler/commands.cpp [hunk 3]
```diff
@@ -1741,23 +1741,23 @@ static void ReverseRangeDimensionsForKernel(NDRDescT &NDR) {
...
+
+  auto setFunc = [&Plugin, Kernel, &DeviceImageImpl, &getMemAllocationFunc,
+                  &Queue](detail::ArgDesc &Arg, size_t NextTrueIndex) {
     switch (Arg.MType) {
     case kernel_param_kind_t::kind_stream:
       break;
     case kernel_param_kind_t::kind_accessor: {
       Requirement *Req = (Requirement *)(Arg.MPtr);
-      AllocaCommandBase *AllocaCmd = getAllocaForReq(Req);
-      RT::PiMem MemArg = (RT::PiMem)AllocaCmd->getMemAllocation();
```
Comment (romanovvlad) [inline sycl/source/detail/scheduler/commands.cpp:1760]: Could you please add an assert checking that if we are here then `getMemAllocationFunc` is not `nullptr`?

File: sycl/source/handler.cpp [hunk 4]
```diff
@@ -132,8 +136,47 @@ event handler::finalize() {
...
+  if (type == detail::CG::Kernel &&
+      MRequirements.size() + MEvents.size() == 0) {
+    // Stream's flush buffer memory is mainly initialized in stream's __init
+    // method. However, this method is not available on host device.
+    // Initializing stream's flush buffer on the host side in a separate task.
+    if (MQueue->is_host()) {
+      for (const detail::StreamImplPtr &Stream : MStreamStorage) {
+        initStream(Stream, MQueue);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (alexanderfle) [reply sycl/source/handler.cpp]: You are right. I will remove this.

File: sycl/source/handler.cpp (hunk 1 above)
Comment (alexanderfle) [reply sycl/source/handler.cpp:137]: Done.

File: sycl/source/detail/scheduler/commands.cpp (hunk 2 above)
Comment (alexanderfle) [reply sycl/source/detail/scheduler/commands.cpp]: Applied the changes.

**Comment by alexanderfle** [comment]
Adding new tests is not expected in the PR, since this is just refactoring of the internal logic.

I have verified that there are already tests that follow this faster way.
Currently, 3 tests follow the way of not changing the graph: `basic_tests/parallel_for_range_host.cpp`, `basic_tests/built-ins.cpp`, `regression/builtins_vector1.cpp`. From llvm-test-suite, more than half of the tests from ESIMD, InlineAsm, InorderQueue, USM, and some others follow this way.

**Review by romanovvlad — APPROVED** [review]

**Pull request merged**